### DISCUSSION TITLE
MacOS menu bar icon dark/light switching

### DISCUSSION THREAD

**Discussion by pixsperdavid** [AvaloniaUI/Avalonia#12954] [Q&A] [answered]
Currently working on a cross-platform utility which runs in the background, with a system tray icon on Windows and a menu bar icon on MacOS.

On MacOS, the dark/light mode of the menu bar itself is determined by the system desktop wallpaper brightness, and is independent of the user-selected dark/light mode for the system. Is there a way to retrieve this property from the system so I can appropriately select a white/black icon?

Alternatively (as detailed here https://bjango.com/articles/designingmenubarextras/), a 'template' image can be supplied which is coloured appropriately by the system. Is this supported by Avalonia?

**Comment by maxkatz6** [comment]
> On MacOS, the dark/light mode of the menu bar itself is determined by the system desktop wallpaper brightness

It is possible to define different bitmap resources for Dark and Light application theme variants (inheriting system variants by default).
But there is no way to specify a resource that depends on background image.

Smart image templates are not supported, but this scenario should be considered with https://github.com/AvaloniaUI/Avalonia/pull/12107 PR. That API from the link reminds me [UWP ShowAsMonocrome](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.bitmapicon.showasmonochrome?view=winrt-22621#windows-ui-xaml-controls-bitmapicon-showasmonochrome).

Aside from using template images, is it possible on macOS to provide multiple icon image sources, so OS can choose by itself when to use which one?

**Comment by pixsperdavid** [comment]
> But there is no way to specify a resource that depends on background image.

This isn't actually necessary for Avalonia to do as the OS makes this determination. Presumably there is an Mac OS SDK method which returns the menu bar dark/light state.

https://developer.apple.com/documentation/appkit/nsimage/1520017-template
According to the above, a bool property needs to be set on an NSImage to indicate that it's a template image. This feels like the best solution to me, although not sure how it would be best expressed in Avalonia.

**Comment by pixsperdavid** [comment, marked as answer]
Resolved by #14348
